File: Problem Solving Pardigms/Divide and Conquer/Binary Search the answer/UVA10341_Solve_it.py
import sys
from math import *


def rl():
    return sys.stdin.readline().strip()


def rn():
    return list(map(int, sys.stdin.readline().strip().split()))


def rln(n):
    l = [None] * n
    for i in range(n):
        l[i] = int(rl())
    return l


def continuous_binary_search(f, lo, hi, gap=1e-5):
    while hi - lo > gap:
        mid = (lo + hi) / 2.0
        if f(mid) < 0:
            hi = mid
        else:
            lo = mid
    return lo


def func(x):
    global p, q, r, s, t, u
    return p * exp(-x) + q * sin(x) + r * cos(x) + s * tan(x) + t * x * x + u


def closest(val):
    center = abs(func(val))
    center_ex = abs(func(val + 0.0001))
    center_def = abs(func(val - 0.0001))
    if center < min(center_def, center_ex):
        return val
    elif center_ex < min(center, center_def):
        return val + 0.0001
    elif center_def < min(center, center_def):
        return val - 0.0001
    else:
        return val


if __name__ == '__main__':
    for line in sys.stdin:
        p, q, r, s, t, u = map(int, line.split())
        if p * exp(-1) + q * sin(1) + r * cos(1) + s * tan(1) + t + u > 1e-9 or p + r + u < 0:
            print('No solution')
            continue

        lo = 0.0
        hi = 1.0
        res = continuous_binary_search(func, lo, hi)
        res = closest(round(res, 4))
        if abs(func(res)) <= 10 ** -2:
            print('{:.4f}'.format(res))
        else:
            print('No solution')

File: Problem Solving Pardigms/Divide and Conquer/Binary Search the answer/UVA_10341_Solve_it.cpp
#include <bits/stdc++.h>

using namespace std;

typedef unsigned int u;

// Math
#define max(a, b) (a > b ? a : b)
#define min(a, b) (a < b ? a : b)

// Loop
#define rep(i, n) for (i = 0; i < n; ++i)
#define repn(i, k, n) for (i = k; i <= n; ++i)
#define repr(i, k, n) for (i = k; i >= n; --i)

// INF
#define INF 1000000000

typedef long long ll;
typedef pair<int, int> ii;
typedef vector<ii> vii;
typedef vector<int> vi;
typedef vector<long long> vll;


template<typename T>
void printList(vector<T> text) {

    for (const auto &value: text)
        std::cout << value << " ";
    std::cout << '\n';
}


double function_imp(int p, int q, int r, int s, int t, int u, double x) {
    double res = p * exp(-x) + q * sin(x) + r * cos(x) + s * tan(x) + t * x * x + u;
    return res;
}


int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w+", stdout);
    int p, q, r, s, t, u;
    double left(0);
    double right(0);
    float val_mid;
    double mid(0);
    while (cin >> p >> q >> r >> s >> t >> u) {

        if (p * exp(-1) + q * sin(1) + r * cos(1) + s * tan(1) + t + u > 1e-9 || p + r + u < 0) {
            cout << "No solution\n";
            continue;
        }

        left = 0.0;
        right = 1.0;
        for (int i = 0; i < 30; ++i) {
            mid = (left + right) / 2;
            val_mid = function_imp(p, q, r, s, t, u, mid);
            if (val_mid > 0) {
                left = mid;

            } else {
                right = mid;
            }
        }
        cout << std::fixed << std::setprecision(4) << left << '\n';
    }


    return 0;
}




